runtime/executor/src/main/java/edu/snu/nemo/runtime/executor/TaskGroupExecutor.java

```diff
@@ -145,303 +201,504 @@ private void initializeDataTransfer() {
         }
-      } catch (final InterruptedException e) {
+        try {
+          encodedBlockSize += sideInputIterator.getNumEncodedBytes();
+        } catch (final DataUtil.IteratorWithNumBytes.NumBytesNotSupportedException e) {
+          encodedBlockSize = -1;
+        }
+
+        LOG.info("log: {} {} read sideInput from InputReader {}",
+            taskGroupId, getPhysicalTaskId(task.getId()), sideInput);
+      } catch (final InterruptedException | ExecutionException e) {
         throw new BlockFetchException(e);
       }
+    });
+  }
+
+  private void sideInputFromThisStage(final Task task, final Map<Transform, Object> sideInputMap) {
+    final String physicalTaskId = getPhysicalTaskId(task.getId());
+
+    taskToInputPipesMap.get(task).forEach(inputPipe -> {
+      if (inputPipe.hasSideInputFor(physicalTaskId)) {
+        // because sideInput is only 1 element in the pipe
+        Object sideInput = inputPipe.remove();
+        final RuntimeEdge inEdge = inputPipe.getSideInputRuntimeEdge();
+        final Transform srcTransform;
+        if (inEdge instanceof PhysicalStageEdge) {
+          srcTransform = ((OperatorVertex) ((PhysicalStageEdge) inEdge).getSrcVertex()).getTransform();
+        } else {
+          srcTransform = ((OperatorTask) inEdge.getSrc()).getTransform();
+        }
+        sideInputMap.put(srcTransform, sideInput);
+        LOG.info("log: {} {} read sideInput from InputPipe {}", taskGroupId, physicalTaskId, sideInput);
+      }
+    });
+  }
+
+  private void prepareTransform(final Transform transform, final Task task) {
+    if (!preparedTransforms.contains(transform)) {
+      final Map<Transform, Object> sideInputMap = new HashMap<>();
+
+      // Check and collect side inputs.
+      if (taskToSideInputReadersMap.keySet().contains(task)) {
+        sideInputFromOtherStages(task, sideInputMap);
+      }
+      if (hasInputPipe(task)) {
+        sideInputFromThisStage(task, sideInputMap);
+      }
 
-      // Check whether there is any output data from the transform and write the output of this task to the writer.
-      final List output = outputCollector.collectOutputList();
-      if (!output.isEmpty() && physicalTaskIdToOutputWriterMap.containsKey(physicalTaskId)) {
-        final long writeStartTime = System.currentTimeMillis();
-        physicalTaskIdToOutputWriterMap.get(physicalTaskId).forEach(outputWriter -> outputWriter.write(output));
-        final long writeEndTime = System.currentTimeMillis();
-        accumulatedWriteTime += writeEndTime - writeStartTime;
-      } // If else, this is a sink task.
+      final Transform.Context transformContext = new ContextImpl(sideInputMap);
+      final LocalPipe outputPipe = taskToOutputPipeMap.get(task);
+      transform.prepare(transformContext, outputPipe);
+
+      preparedTransforms.add(transform);
+    }
+  }
+
+  /**
+   * Executes the task group.
+   */
+  public void execute() {
+    final Map<String, Object> metric = new HashMap<>();
```

Comment:
Please refactor metric collection code into a separate class, so that its LOC is minimal in `TaskGroupExecutor`.